### ISSUE TITLE
DataFrame Column Issues

### ISSUE THREAD

**Issue by brittainhard** [open-source-economics/Tax-Calculator#1367]
@martinholmer @MattHJensen @PeterDSteinberg since the API for `dropq` has been refactored, I am getting DataFrame errors with the new function.

The error is `Error: Length mismatch: Expected axis has 219813 elements, new values have 219814 elements`.

It seems like it might be a quick fix, but I'm not sure where to look. Any ideas?

**Comment by PeterDSteinberg** [comment]
@brittainhard  What's the full traceback?  Maybe it is a pandas update thing, like handling edge cases on joins with nulls or something like that?  The B-Tax issue we have resolved in last 24 hours was a pandas internal change that we didn't test before our dev release.  One way to test it out would be to install using `python setup.py develop` on each of the tagged commits from the repos, then install the prereqs modified to have the older numpy / pandas - may help isolate the pandas code issues vs our own.

**Comment by PeterDSteinberg** [comment]
@brittainhard  I think this problem may be related to this `chooser` function: https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/dropq/dropq_utils.py#L198

Or the usage of `transform` with `chooser` as the argument.  I can look into it further after my meeting this morning.

**Comment by PeterDSteinberg** [comment]
This bug may relate to pandas updates and how `pd.cut` works with bins and including which edge of each bin (right edge or not).  Or it may have to do with the `transform` function `chooser` returning an unexpected data type?  (something in `taxcalc.dropq.dropq_utils` related to pandas updates)

```
    2017-05-17 00:15:46,056 DEBG 'celery' stderr output:
    [2017-05-17 00:15:46,056: WARNING/MainProcess] {'start_year': 2017, 'user_mods': {'growdiff_response': {}, 'consumption': {}, 'growdiff_baseline': {}, 'behavior': {}, 'policy': {2017: {u'_NIIT_PT_taxed': [False], u'_FICA_ss_trt': [0.2], u'_ID_BenefitCap_Switch': [[True, True, True, True, 1.0, 1.0, 1.0]], u'_ALD_InvInc_ec_base_RyanBrady': [False], u'_EITC_indiv': [False], u'_ID_BenefitSurtax_Switch': [[True, True, True, True, True, True, True]], u'_CTC_new_refund_limited': [False], u'_CG_nodiff': [False]}}, 'gdp_elasticity': {}}, 'year_n': 1}
    
    2017-05-17 00:15:46,664 DEBG 'flask' stderr output:
    127.0.0.1 - - [17/May/2017 00:15:46] "GET /dropq_query_result?job_id=be87a13a-a68c-474a-bb7f-cf7ddd8c8a09 HTTP/1.0" 200 -
    
    2017-05-17 00:15:46,674 DEBG 'flask' stderr output:
    --
    
    2017-05-17 00:17:55,991 DEBG 'flask' stderr output:
    127.0.0.1 - - [17/May/2017 00:17:55] "GET /dropq_query_result?job_id=4cdd7704-5b10-49a3-84ae-c4ac840bc4ea HTTP/1.0" 200 -
    
    2017-05-17 00:17:56,204 DEBG 'celery' stderr output:
    [2017-05-17 00:17:56,203: ERROR/MainProcess] Task celery_tasks.dropq_task_async[4cdd7704-5b10-49a3-84ae-c4ac840bc4ea] raised unexpected: ValueError('Length mismatch: Expected axis has 219813 elements, new values have 219814 elements',)
    Traceback (most recent call last):
      File "/home/ubuntu/miniconda2/envs/aei_dropq/lib/python2.7/site-packages/celery/app/trace.py", line 367, in trace_task
        R = retval = fun(*args, **kwargs)
      File "/home/ubuntu/miniconda2/envs/aei_dropq/lib/python2.7/site-packages/celery/app/trace.py", line 622, in __protected_call__
        return self.run(*args, **kwargs)
      File "/home/ubuntu/deploy/celery_tasks.py", line 144, in dropq_task_async
        return dropq_task(year, user_mods, first_budget_year, beh_params, tax_dta_full)
      File "/home/ubuntu/deploy/celery_tasks.py", line 121, in dropq_task
        fiscal_tot_i_bl, fiscal_tot_i_ref) = taxcalc.dropq.run_nth_year_tax_calc_model(**kw)
      File "/home/ubuntu/miniconda2/envs/aei_dropq/lib/python2.7/site-packages/taxcalc/dropq/dropq.py", line 70, in run_nth_year_tax_calc_model
        itax_sum2, ptax_sum2, comb_sum2) = dropq_summary(rawres1, rawres2, mask)
      File "/home/ubuntu/miniconda2/envs/aei_dropq/lib/python2.7/site-packages/taxcalc/dropq/dropq_utils.py", line 293, in dropq_summary
        df1, df2 = drop_records(df1, df2, mask)
      File "/home/ubuntu/miniconda2/envs/aei_dropq/lib/python2.7/site-packages/taxcalc/dropq/dropq_utils.py", line 257, in drop_records
        df2['flag_dec'] = gp2_dec['mask'].transform(chooser)
      File "/home/ubuntu/miniconda2/envs/aei_dropq/lib/python2.7/site-packages/pandas/core/groupby.py", line 2990, in transform
        result.index = self._selected_obj.index
      File "/home/ubuntu/miniconda2/envs/aei_dropq/lib/python2.7/site-packages/pandas/core/generic.py", line 2983, in __setattr__
        return object.__setattr__(self, name, value)
      File "pandas/_libs/src/properties.pyx", line 65, in pandas._libs.lib.AxisProperty.__set__ (pandas/_libs/lib.c:45103)
      File "/home/ubuntu/miniconda2/envs/aei_dropq/lib/python2.7/site-packages/pandas/core/series.py", line 308, in _set_axis
        self._data.set_axis(axis, labels)
      File "/home/ubuntu/miniconda2/envs/aei_dropq/lib/python2.7/site-packages/pandas/core/internals.py", line 2836, in set_axis
        (old_len, new_len))
    ValueError: Length mismatch: Expected axis has 219813 elements, new values have 219814 elements
```

cc @martinholmer @brittainhard @MattHJensen

**Comment by martinholmer** [comment]
@brittainhard said seventeen days ago:
> since the API for dropq has been refactored, I am getting DataFrame errors with the new function.

Yes, it is possible I made a mistake when I refactored the dropq code.  The refactored functions passed all the tests written by @talumbau, but perhaps that dropq test suite is not strong enough to catch all errors.

Can you provide more information about where you think the error is in the refactored dropq code?  For example, which dropq function were you calling when you encountered the error.

I'd be happy to fix the dropq code but need from you additional test code that fails because of the bug.  Once I have that additional test, I can revise the dropq code to pass that test and all the pre-existing tests.

I'm looking forward to your bug report.

@PeterDSteinberg

**Comment by PeterDSteinberg** [comment]
@martinholmer This is the code that led to the error quoted above:

```
import glob
from taxcalc import dropq_utils
from taxpuf import get_puf
import pandas as pd

df = get_puf()
from taxcalc.dropq import run_nth_year_tax_calc_model
c = {'start_year': 2017, 'user_mods': {'growdiff_response': {}, 'consumption': {}, 'growdiff_baseline': {}, 'behavior': {}, 'policy': {2017: {u'_NIIT_PT_taxed': [False], u'_FICA_ss_trt': [0.2], u'_ID_BenefitCap_Switch': [[True, True, True, True, 1.0, 1.0, 1.0]], u'_ALD_InvInc_ec_base_RyanBrady': [False], u'_EITC_indiv': [False], u'_ID_BenefitSurtax_Switch': [[True, True, True, True, True, True, True]], u'_CTC_new_refund_limited': [False], u'_CG_nodiff': [False]}}, 'gdp_elasticity': {}}, 'year_n': 9}
out = run_nth_year_tax_calc_model(9, 2017, df, c['user_mods'], return_json=True)



```

**Comment by martinholmer** [comment]
Issue #1367 has been resolved by merge of pull request #1387.
